src/middleware/mod.rs

```diff
@@ -1,13 +1,15 @@
+use std::sync::Arc;
+
 use futures::future::FutureObj;
 
-use crate::{head::Head, Request, Response, RouteMatch};
+use crate::{endpoint::BoxedEndpoint, head::Head, Request, Response, RouteMatch};
 
 mod default_headers;
 
 pub use self::default_headers::DefaultHeaders;
 
 /// Middleware for an app with state of type `Data`
-pub trait Middleware<Data>: Send + Sync {
+pub trait ReqResMiddleware<Data>: Send + Sync {
```

Comment:
I think we can probably drop this trait for now -- unless you think there's a strong need for it?

Reply:
If we do keep it, we could consider providing a blanket impl rather than an adapter:

```rust
impl<M, Data> Middleware<Data> for M where M: ReqResMiddleware<Data> { ... }
```

Reply:
I've tried to add a blanked impl, but it made a conflict with impl for `DefaultHeaders`.

Reply:
Ahha, makes sense! Do you want to just nix this trait for the time being?

Reply:
Done!